### ISSUE TITLE
dkms-new-kernel.sh is applied to newer native kernels (default boot kernel), but not newer lts kernels

### ISSUE THREAD

**Issue by marioroy** [clearlinux/distribution#2829]
Given the following scenario, dkms modules are not installed for the lts kernel.

1. Install kernel-lts
2. Install kernel-native and make it the default boot kernel
3. Simulate periodic updates over time

What happens is that the dkms-new-kernel.sh is applied to newer native kernels, but not newer lts kernels.

Ditto, vice-versa running with the lts kernel as default. dkms-new-kernel.sh is applied to newer lts kernels, but not native.

**Comment by marioroy** [comment]
Ping

**Comment by bwarden** [comment]
See https://github.com/clearlinux-pkgs/dkms/blob/main/0001-Add-script-to-run-autoinstall-on-new-kernel.patch

My original design was to only run `dkms autoinstall` for the new kernel you're most likely going to boot next. Of course at any time you can run `dkms autoinstall -k <kernel_ver>` to manually rebuild DKMS modules for any installed kernel.

Do you often switch between LTS and native kernels? I could rewrite it to always run for all the default kernels, but since this is triggered by `swupd update`, I didn't want to extend that runtime unnecessarily.

**Comment by marioroy** [comment]
> Do you often switch between LTS and native kernels?

Typically, not. The LTS kernel is there in the event the native kernel fails for an unknown reason. Having dkms run automatically is what most Linux distributions do, no matter having one or two kernel installed. 

> I didn't want to extend that runtime unnecessarily.

I have a script `check-kernel-dkms`, included with the nvidia-driver-on-clear-linux repo. Please account for the fact that this is checking both kernels.

```bash
$ time ./check-kernel-dkms 
Double-checking dkms modules in 5.15.93-37.lts2021.
Double-checking dkms modules in 6.1.11-1261.native.

real   0m0.202s
user   0m0.134s
sys    0m0.092s
```

Ah, I see. Is swupd update running dkms autoinstall every time, regardless if there was a kernel update. At the very least, dkms should run if installing/updating a given kernel. CL user(s) ensure the well-being of the kernels on the system. Only for swupd update to break it :( Breaking the LTS kernel, simply because omitting the dkms step is not fun.

_The swupd process is all over the native kernel by ensuring its well-being. And then, could care less about the other kernel. I just want to scream, but instead reminded of the Serenity Prayer._

**Comment by bwarden** [comment]
The way the script currently works -- it checks which variant of kernel you are currently running (native, lts, etc.), then looks for the default symlink _for that variant_, e.g. `/usr/lib/kernel/default-native`. If that symlink points to a different version _of that variant_ than you're currently running, it calls `dkms autoupdate`. That way it only runs if a _new_ kernel _of the same variant you're currently running_ is installed.

Granted, if you run it against a kernel for which it has already built the modules, `dkms autoupdate` will just return quickly, so I could either:
- run it against each kernel pointed to by `/usr/lib/kernel/default-*` (so only the latest kernel per variant), or
- run it against each version (2-3) of each kernel variant

Again, I'm trying not to make `swupd update` take any more time than it really needs to, so I'm inclined to do the first, unless there's a really good reason. I've sat through ridiculous amounts of time on rpm-based systems waiting for DKMS to build graphics driver modules for every installed kernel, and I don't want to invoke that here.

I'll leave discussion of swupd's handling of multiple versions and variants in #2822.

**Comment by marioroy** [comment]
> so only the latest kernel per variant

That would be awesome. only the latest per variant.

**Comment by marioroy** [comment]
> Granted, if you run it against a kernel for which it has already built the modules, dkms autoupdate will just return quickly.

This is true, like shown above in the captured output.

**Comment by bwarden** [comment]
OK, the fix is in the pipeline now. I also found a bug in dkms-remove-old.sh, preventing it from actually removing old modules.

**Comment by marioroy** [comment]
> I also found a bug in dkms-remove-old.sh, preventing it from actually removing old modules.

That was going to be my next issue ticket (part 3 of 3 in the dkms series). Thank you for fixing dkms-remove-old.sh. I've been clearing the uninstalled kernel bits for some time in /var/lib/dkms/.

**Comment by bwarden** [comment]
Next I think I'm going to move the installed modules to `/lib/modules/<kernel>/updates/dkms/` -- that way `swupd repair` will leave them alone, but they should still be reaped by clr-boot-manager when it deletes old kernels, and if not that, dkms-remove-old.sh should clean it up too.

**Comment by bwarden** [comment]
`sudo journalctl -u dkms-\*` would show what happened there. `dkms-remove-old.sh` should have called `dkms remove` for all kernels that are no longer present, and you can re-run it manually anytime, too. I had out-of-tree modules (ixgbe) installed on my system while these DKMS changes applied overnight and it all went smoothly, so I'm not sure what went wrong on your system.

clr-boot-manager changes to clean up non-selected kernels are being tracked in that other issue, still.

**Comment by bwarden** [comment]
```
$ ls -1 /lib/modules
5.15.93-37.lts2021
5.15.94-38.lts2021
6.0.2-1201.native
6.1.11-1261.native
6.1.12-1265.native
$ ls -1 /var/lib/dkms/ixgbe/
5.18.11
kernel-5.15.94-38.lts2021-x86_64
kernel-6.0.2-1201.native-x86_64
kernel-6.1.12-1265.native-x86_64
original_module
$ ls -1 /var/lib/dkms/ixgbe/5.18.11/
5.15.94-38.lts2021
6.0.2-1201.native
6.1.12-1265.native
source
```
The leftover module in these two kernels is from the pre-modified DKMS archiving/restoring the original version of the module. I eliminated that in the current implementation so this shouldn't happen going forward, but we still obviously need to cleanup the rest of these module paths:
```
$ find /lib/modules/5.15.93-37.lts2021/ -type f
/lib/modules/5.15.93-37.lts2021/modules.builtin.bin
/lib/modules/5.15.93-37.lts2021/modules.devname
/lib/modules/5.15.93-37.lts2021/modules.softdep
/lib/modules/5.15.93-37.lts2021/kernel/drivers/net/ethernet/intel/ixgbe/ixgbe.ko
/lib/modules/5.15.93-37.lts2021/modules.builtin.alias.bin
/lib/modules/5.15.93-37.lts2021/modules.symbols
/lib/modules/5.15.93-37.lts2021/modules.dep.bin
/lib/modules/5.15.93-37.lts2021/modules.alias
/lib/modules/5.15.93-37.lts2021/modules.symbols.bin
/lib/modules/5.15.93-37.lts2021/modules.dep
/lib/modules/5.15.93-37.lts2021/modules.alias.bin
$ find /lib/modules/6.1.11-1261.native/ -type f
/lib/modules/6.1.11-1261.native/modules.builtin.bin
/lib/modules/6.1.11-1261.native/modules.devname
/lib/modules/6.1.11-1261.native/modules.softdep
/lib/modules/6.1.11-1261.native/kernel/drivers/net/ethernet/intel/ixgbe/ixgbe.ko
/lib/modules/6.1.11-1261.native/modules.builtin.alias.bin
/lib/modules/6.1.11-1261.native/modules.symbols
/lib/modules/6.1.11-1261.native/modules.dep.bin
/lib/modules/6.1.11-1261.native/modules.alias
/lib/modules/6.1.11-1261.native/modules.symbols.bin
/lib/modules/6.1.11-1261.native/modules.dep
/lib/modules/6.1.11-1261.native/modules.alias.bin
```

**Comment by marioroy** [comment]
I did not have out-of-tree (very old versions). Quite simply, I ran `swupd update` and captured ls output. I needed to run three commands to clear up the older kernel version (lts) not removed automatically and dkms artifacts for both variants.

**Comment by marioroy** [comment]
Are users expected to run these commands manually, which is automatic in other Linux distributions (e.g. Fedora Linux, Ubuntu Linux)?

1. Installing a new or updating a kernel. Run `dkms autoinstall`. This is working :)
2. Removing a kernel from the system. The `dkms remove` should come first before removing the kernel from the system.

Interestingly, running `clr-boot-manager update` removes the older kernel version for the default kernel. But requires manual intervention getting the older lts or other kernel removed. Otherwise, the boot partition will fill up.

**Comment by bwarden** [comment]
> I did not have out-of-tree (very old versions). Quite simply, I ran `swupd update` and captured ls output. I needed to run three commands to clear up the older kernel version (lts) not removed automatically and dkms artifacts for both variants.

> Are users expected to run these commands manually, which is automatic in other Linux distributions (e.g. Fedora Linux, Ubuntu Linux)?

OK, I get it. Right -- dkms-remove-old.sh is only triggered by swupd update, not by clr-boot-manager, so those dkms entries will stick around until swupd update does something again, even though you manually removed the kernels via clr-boot-manager.

We'll think about whether clr-boot-manager should trigger those jobs too (`dkms-new-kernel.service` and `dkms-remove-old.service`), but that will take some de-clunking, as swupd update calls clr-boot-manager internally.

In a nutshell, today:
- `dkms autoinstall` is run for each installed kernel every time swupd update actually updates _anything_, via `dkms-new-kernel.service` running `dkms-new-kernel.sh`.
- `dkms remove` is run for each kernel that is no longer installed, every time swupd update actually updates _anything_, via `dkms-remove-old.service` running `dkms-remove-old.sh`.
- Neither `clr-boot-manager update` nor `clr-boot-manager remove-kernel <>` triggers these jobs.

So, any installed DKMS modules should be automatically built for any new kernel that's added by swupd, and they'll _eventually_ be removed for any kernel that's removed, just not until swupd update triggers that job again.

**Comment by marioroy** [comment]
You have no idea on how grateful I am pertaining to `dkms autoinstall` working 100%.

For the other side of the coin `dkms remove`, I find it interesting of the inconsistencies. Well, hats off to the progress already made. I hope for clr-boot-manager, dkms-remove-old.service, or whatever tool to be consistent and not miss the dkms bits when uninstalling a kernel.

Blessings and grace.
